ImperatorToCK3/Source/CK3/CK3World.cpp

```diff
@@ -245,22 +250,110 @@ void CK3::World::linkCountiesToTitleHolders(const ImperatorWorld::World& sourceW
 			auto countyTitle = std::make_shared<Title>();
 			countyTitle->titleName = name;
 			const auto capitalBaronyProvince = landedTitle.capitalBaronyProvince;
+
 			if (capitalBaronyProvince >0) // 0 is not a valid province in CK3
 			{
-				auto owner = provinces.find(capitalBaronyProvince)->second->getOwner();
-				if (owner != "0")
+				if (provinces.find(capitalBaronyProvince)==provinces.end())
+					LOG(LogLevel::Warning) << "Capital barony province not found " << capitalBaronyProvince;
+				else
 				{
-					auto impMonarch = -1;
-					if (sourceWorld.getCountries().find(stoi(owner)) != sourceWorld.getCountries().end()) impMonarch = sourceWorld.getCountries().find(stoi(owner))->second->getMonarch();
-					if (impMonarch>=0) countyTitle->holder = std::to_string(impMonarch);
+					auto impProvince = provinces.find(capitalBaronyProvince)->second->srcProvince;
+					if (impProvince)
+					{
+						auto impMonarch = -1;
+						if (impWorld.getCountries().find(impProvince->getOwner()) != impWorld.getCountries().end()) impMonarch = impWorld.getCountries().find(impProvince->getOwner())->second->getMonarch();
+						if (impMonarch >= 0) countyTitle->holder = "imperator" + std::to_string(impMonarch);
+					}
+					else // county is probably outside of Imperator map
+					{
+						auto vanillaHistory = titlesHistory.popTitleHistory(name);
+						if (vanillaHistory) countyTitle->historyString = *vanillaHistory;
+					}
 				}
-				else // county is probably outside of Imperator map
+			}
+			titles.insert(std::pair(name, countyTitle));
+		}
+	}
+}
+
+
+void CK3::World::insertVanillaNonCountiesToTitles(const ImperatorWorld::World& impWorld)
+{
+	for (const auto& [name, landedTitle] : landedTitles.getFoundTitles())
+	{
+		if (name.find("c_") != 0 && name.find("b_") != 0 ) // title is a duchy or higher
+		{
+			bool toInsert = true;
+			// important check: if any of the title's de jure counties's holder is "0", don't insert the title
+			for (const auto& [vassalTitleName, deJureVassal] : landedTitle.getFoundTitles())
```

Comment:
this is O(N^2). Never do this. Make a cache and then poke at the cache for 2N.

Reply:
Not pure O(n^2) in reality. The second loop iterates only though `landedTitle`'s de jure vassals, and only if `landedTitle` is above county level.
How should I make a cache (I'm not very familiar with caches)?